src/puppetlabs/puppetdb/queue.clj

```diff
@@ -0,0 +1,199 @@
+(def parse-metadata (metadata-parser))
+
+(def parse-cmd-filename
+  (let [parse-metadata (metadata-parser (cons "unknown" metadata-command-names))]
+    (fn [s]
+      (let [rx #"([0-9]+)-(.*)"]
+        (when-let [[_ id qmeta] (re-matches rx s)]
+          (parse-metadata qmeta))))))
+
+(defrecord CommandRef [id command version certname received callback annotations delete?])
+
+(defn cmdref->entry [{:keys [id command version certname received]}]
+  (stock/entry id (metadata-str received command version certname)))
+
+(defn entry->cmdref [entry]
+  (let [{:keys [received command version certname]} (-> entry
+                                                        stock/entry-meta
+                                                        parse-metadata)]
+    (map->CommandRef {:id (stock/entry-id entry)
+                      :command command
+                      :version version
+                      :certname certname
+                      :received received
+                      :callback identity
+                      :annotations {:id (stock/entry-id entry)
+                                    :received received}})))
+
+(defn cmdref->cmd [q cmdref]
+  (let [entry (cmdref->entry cmdref)]
+    (with-open [command-stream (stock/stream q entry)]
+      (assoc cmdref
+             :payload (stream->json command-stream)
+             :entry entry))))
+
+(defn cons-attempt [cmdref exception]
+  (update cmdref :attempts conj {:exception exception
+                                 :time (kitchensink/timestamp)}))
+
+(defn store-command
+  ([q command version certname command-stream]
+   (store-command q command version certname command-stream identity))
+  ([q command version certname command-stream command-callback]
+   (let [current-time (time/now)
+         entry (stock/store q
+                            command-stream
+                            (metadata-str current-time command version certname))]
+     (map->CommandRef {:id (stock/entry-id entry)
+                       :command command
+                       :version version
+                       :certname certname
+                       :callback command-callback
+                       :received (kitchensink/timestamp current-time)
+                       :annotations {:id (stock/entry-id entry)
+                                     :received (kitchensink/timestamp current-time)}}))))
+
+(defn ack-command
+  [q command]
+  (stock/discard q (:entry command)))
+
+(deftype SortedCommandBuffer [^TreeMap fifo-queue ^HashMap certnames-map ^long max-entries ^clojure.lang.IFn delete-update-fn]
+  async-protos/Buffer
+  (full? [this]
+    (>= (.size fifo-queue) max-entries))
+
+  (remove! [this]
+    (let [^CommandRef cmdref (val (.pollFirstEntry fifo-queue))
+          command-type (:command cmdref)]
+      (when (or (= command-type "replace catalog")
+                (= command-type "replace facts"))
+        (.remove certnames-map [command-type (:certname cmdref)]))
+      cmdref))
```

Comment:
nils are never ok in core.async. If there's nothing there, it will block on take until something is AFAIK